src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -0,0 +1,668 @@
+-- 'vertexCount' (overlay x y) <= 'vertexCount' x + 'vertexCount' y
+-- 'edgeCount'   (overlay x y) >= 'edgeCount' x
+-- 'edgeCount'   (overlay x y) <= 'edgeCount' x   + 'edgeCount' y
+-- @
+overlay :: (Ord a, Ord b) => AdjacencyMap a b -> AdjacencyMap a b -> AdjacencyMap a b
+overlay (BAM lr1 rl1) (BAM lr2 rl2) = BAM (Map.unionWith Set.union lr1 lr2) (Map.unionWith Set.union rl1 rl2)
+
+-- | /Connect/ two bipartite graphs, not adding the edges between vertices in
+-- the same part. This is a commutative and associative operation with the
+-- identity 'empty', which distributes over 'overlay' and obeys the
+-- decomposition axiom.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory. Note that the
+-- number of edges in the resulting graph is quadratic with respect to the
+-- number of vertices in the arguments: /m = O(m1 + m2 + l1 * r2 + l2 * r1)/.
+--
+-- @
+-- connect ('leftVertex' 1) ('rightVertex' "a")      == 'edge' 1 "a"
+-- connect ('leftVertex' 1) ('rightVertex' 1)        == 'edge' 1 1
+-- connect ('leftVertex' 1) ('leftVertex' 2)         == 'vertices' [1, 2] []
+-- connect ('vertices' [1] [4]) ('vertices' [2] [3]) == 'edges' [(1, 3), (2, 4)]
+-- 'isEmpty'     (connect x y)                     == 'isEmpty'   x   && 'isEmpty'   y
+-- 'hasVertex' z (connect x y)                     == 'hasVertex' z x || 'hasVertex' z y
+-- 'vertexCount' (connect x y)                     >= 'vertexCount' x
+-- 'vertexCount' (connect x y)                     <= 'vertexCount' x + 'vertexCount' y
+-- 'edgeCount'   (connect x y)                     >= 'edgeCount' x
+-- 'edgeCount'   (connect x y)                     >= 'edgeCount' y
+-- 'edgeCount'   (connect x y)                     >= 'leftVertexCount' x * 'rightVertexCount' y
+-- 'edgeCount'   (connect x y)                     <= 'leftVertexCount' x * 'rightVertexCount' y + 'rightVertexCount' x * 'leftVertexCount' y + 'edgeCount' x + 'edgeCount' y
+-- @
+connect :: (Ord a, Ord b) => AdjacencyMap a b -> AdjacencyMap a b -> AdjacencyMap a b
+connect (BAM lr1 rl1) (BAM lr2 rl2) = BAM lr rl
+    where
+        lr = Map.unionsWith Set.union $
+            [ lr1, lr2
+            , Map.fromSet (const $ Map.keysSet rl2) (Map.keysSet lr1)
+            , Map.fromSet (const $ Map.keysSet rl1) (Map.keysSet lr2)
+            ]
+        rl = Map.unionsWith Set.union $
+            [ rl1, rl2
+            , Map.fromSet (const $ Map.keysSet lr2) (Map.keysSet rl1)
+            , Map.fromSet (const $ Map.keysSet lr1) (Map.keysSet rl2)
+            ]
+
+-- | Construct the graph comprising two given lists of isolated vertices for
+-- each part.
+-- Complexity: /O(L * log(L))/ time and /O(L)/ memory, where /L/ is the total
+-- length of two lists.
+--
+-- @
+-- vertices [] []                               == 'empty'
+-- vertices [x] []                              == 'leftVertex' x
+-- vertices [] [x]                              == 'rightVertex' x
+-- 'hasLeftVertex' x . ('flip' ('const' [])) vertices == 'elem' x
```

Comment:
This and the next examples seem too complex, I suggest simplifying them by dropping point-free notation:

```haskell
hasLeftVertex  x (vertices ys zs) == elem x ys
hasRightVertex x (vertices ys zs) == elem x zs
```